### PR TITLE
Constrain unique version names for workflow versions plus tags

### PR DESCRIPTION
**Pull request by denis-yuen**
A structural fix which is probably controversial for what would normally be considered a hotfix. Getting reviews for discussion

* locally working fix for https://github.com/dockstore/dockstore/issues/3252
* introduces constraints (note that constraints for tags can be foreign keys since they always go with tools, workflowversions are used for both services and bioworkflow, so we can't put in a foreign key constraint .... both get not null constraints and unique constraints though)
* a number of endpoints are changed, when working with DAOs you can no longer create a tag/workflowversion without a valid parent at the time of creation(not null constraints are not deferrable), in some parts of the code we used to create versions and link them up afterwards
* a handful of named queries are changed and two join tables are deleted

Overall, this is a tricky PR that could have side-effects/needs some careful testing (thanks for existing testing for catching a number of edge cases already)

TODO: https://ucsc-cgl.atlassian.net/browse/SEAB-870
may need help with these items
* will need further testing and/or[ better migration](https://github.com/dockstore/dockstore/blob/8b03a7143725f56c9611541026529976ee8bd694/dockstore-webservice/src/main/resources/migrations.1.8.0.xml#L210) to make sure we're not deleting too much when introducing constraints, but also delete source files and other dependent records linked to the now deleted versions
* will need to test adding aliases more carefully (had to update native query, so converted to JPQL while at it)

### CODE REVIEW

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=5050953&bid=16630161)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=5050953&bid=16630161)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=5050953&bid=16630161)

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java [hunk 1]
```diff
@@ -217,11 +217,17 @@ protected void updateDBWorkflowWithSourceControlWorkflow(Workflow workflow, Work
...
-                // create a new one and replace the old one
+                // attach real workflow
+                workflow.addWorkflowVersion(version);
+
                 final long workflowVersionId = workflowVersionDAO.create(version);
                 releaseCreated = true;
                 workflowVersionFromDB = workflowVersionDAO.findById(workflowVersionId);
                 workflow.getWorkflowVersions().add(workflowVersionFromDB);
+                // create a new one and replace the old one if it exists
+                if (existingVersionMap.containsKey(workflowVersionFromDB.getName())) {
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java]: I don't understand the need for this -- on line 214 you've already verified that the map does not contain the name. How could the map now contain the name? Can line 221/223 change the name? Otherwise this seems redundant. Or I'm missing something.

File: dockstore-webservice/src/main/resources/migrations.1.8.0.xml [hunk 2]
```diff
@@ -181,4 +181,85 @@
...
+
+    <changeSet author="dyuen" id="cleanup invalid content">
+        <!-- need to investigate whether too much content is being dropped and whether the right content is being dropped -->
+        <comment>some orphaned versions are protected by security</comment>
+        <sql dbms="postgresql">
+            alter table workflowversion disable row level security;
+            alter table tag disable row level security;
+        </sql>
+
+        <comment>delete the versions that are orphaned already</comment>
```
[the remark this answers is not in the document]
Comment (denis-yuen) [reply dockstore-webservice/src/main/resources/migrations.1.8.0.xml]: I think that would be reasonable to do, I'll add it to the TODO when this spawns new tickets

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java (hunk 1 above)
Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java]: I think it should be redundant, I probably had something else in mind.

File: dockstore-webservice/src/main/resources/migrations.1.8.0.xml [hunk 3]
```diff
@@ -181,4 +181,85 @@
...
+            </column>
+        </addColumn>
+        <!-- migrate data -->
+        <sql dbms="postgresql">
+            update workflowversion set parentid = wwv.workflowid from workflowversion as wv join workflow_workflowversion as wwv on wv.id = wwv.workflowversionid where workflowversion.id = wv.id;
+            update tag set parentid = tt.toolid from tag as t join tool_tag as tt on t.id = tt.tagid where tag.id = t.id;
+        </sql>
+
+        <!-- clean-up -->
+        <dropForeignKeyConstraint baseTableName="tool_tag" constraintName="fkjkn6qubuvn25bun52eqjleyl6"/>
```
[the remark this answers is not in the document]
Comment (denis-yuen) [reply dockstore-webservice/src/main/resources/migrations.1.8.0.xml]: I might not follow, but unfortunately they do appear to have these names in prod

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=5050953&bid=16630161)
